grails-app/services/org/pih/warehouse/inventory/InventorySnapshotService.groovy

```diff
@@ -464,155 +464,207 @@ class InventorySnapshotService {
-        def balanceOpeningMap = getQuantityOnHandByProduct(location, startDate)
-        if (balanceOpeningMap.isEmpty()) {
-            throw new IllegalStateException("No inventory snapshot for ${startDate}")
+    List<TransactionEntry> getCreditTransactionEntries(List<TransactionEntry> transactionEntries) {
+        return transactionEntries.findAll {
+            it.transaction.transactionType.transactionCode == TransactionCode.CREDIT && it.quantity > 0
         }
+    }
 
-        // Get ending balance
-        def balanceClosingMap = getQuantityOnHandByProduct(location, endDate)
-        if (balanceClosingMap.isEmpty()) {
-            throw new IllegalStateException("No inventory snapshot for ${endDate}")
+    List<TransactionEntry> getDebitTransactionEntries(List<TransactionEntry> transactionEntries) {
+        return transactionEntries.findAll {
+            it.transaction.transactionType.transactionCode == TransactionCode.DEBIT ||
+                    (it.transaction.transactionType.transactionCode == TransactionCode.CREDIT && it.quantity < 0)
         }
+    }
 
+    List<TransactionEntry> getFilteredTransactionEntries(
+            List<TransactionCode> transactionCodes,
+            Date startDate,
+            Date endDate,
+            List<Category> categories,
+            List<Tag> tagsList,
+            List<ProductCatalog> catalogsList,
+            Location location,
+            String orderBy
+    ) {
+        return TransactionEntry.createCriteria().list {
+            inventoryItem {
+                product {
+                    if (categories) {
+                        'in'('category', categories)
+                    }
+
+                    if (tagsList) {
+                        tags {
+                            'in'("id", tagsList.id)
+                        }
+                    }
+
+                    if (catalogsList) {
+                        productCatalogItems {
+                            productCatalog {
+                                'in'("id", catalogsList*.id)
+                            }
+                        }
+                    }
+                }
+            }
+            transaction {
+                if (transactionCodes) {
+                    transactionType {
+                        'in'('transactionCode', transactionCodes)
+                    }
+                }
 
-        // We need all products that have either an opening balance or closing balance
-        def products = new HashSet()
-        products.addAll(balanceOpeningMap.keySet())
-        products.addAll(balanceClosingMap.keySet())
-
-        def data = []
-        products.findAll { (categories.contains(it.category) && ((tags && it.hasOneOfTags(tags)) || (catalogs && it.hasOneOfCatalogs(catalogs)))) ||
-                (!tags && !catalogs && categories.contains(it.category)) }.each { Product product ->
+                if (startDate) {
+                    gt('transactionDate', startDate)
```

Comment:
Should one of the dates (start or end) be also equal? `gte` for start date or `lte` for end date? Looking at your diagram in the description, you could miss a transaction that is at the exact date as your opening balance date or closing balance date (if both are excluded). How was this calculated before? Or am I overthinking this part? :thinking: